pkg/resourcemanager/controller/managedresource/reconciler.go

```diff
@@ -164,26 +160,24 @@ func (r *Reconciler) reconcile(ctx context.Context, log logr.Logger, mr *resourc
 	// Initialize condition based on the current status.
 	conditionResourcesApplied := v1beta1helper.GetOrInitConditionWithClock(r.Clock, mr.Status.Conditions, resourcesv1alpha1.ResourcesApplied)
 
-	for _, ref := range mr.Spec.SecretRefs {
-		secret := &corev1.Secret{}
-		if err := r.SourceClient.Get(reconcileCtx, client.ObjectKey{Namespace: mr.Namespace, Name: ref.Name}, secret); err != nil {
+	// fetch all referenced secrets
+	referencedSecrets := make([]*corev1.Secret, len(mr.Spec.SecretRefs))
+	for i, ref := range mr.Spec.SecretRefs {
+		referencedSecrets[i] = &corev1.Secret{}
+		if err := r.SourceClient.Get(reconcileCtx, client.ObjectKey{Namespace: mr.Namespace, Name: ref.Name}, referencedSecrets[i]); err != nil {
 			conditionResourcesApplied = v1beta1helper.UpdatedConditionWithClock(r.Clock, conditionResourcesApplied, gardencorev1beta1.ConditionFalse, "CannotReadSecret", err.Error())
 			if err := updateConditions(ctx, r.SourceClient, mr, conditionResourcesApplied); err != nil {
 				return reconcile.Result{}, fmt.Errorf("could not update the ManagedResource status: %w", err)
 			}
 
-			return reconcile.Result{}, fmt.Errorf("could not read secret '%s': %+v", secret.Name, err)
+			return reconcile.Result{}, fmt.Errorf("could not read secret '%s': %+v", referencedSecrets[i].Name, err)
```

Comment:
```suggestion
                        return reconcile.Result{}, fmt.Errorf("could not read secret '%s': %+v", ref.Name, err)
````

It wasn't changed by this PR but still think it's safer to use `ref.Nane` in case of an error.